src/offset/library.cairo

```diff
@@ -312,307 +277,192 @@ namespace CarbonableOffseter {
-        assert users[index] = user;
-        if (index == 0) {
-            return ();
-        }
-        _read_users_iter(index=index - 1, users=users);
-        return ();
-    }
 
-    func _registerd_tokens_iter{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-        contract_address: felt, user: felt, index: felt, tokens_len: felt, tokens: Uint256*
-    ) -> (tokens_len: felt, tokens: Uint256*) {
-        alloc_locals;
-
-        // Get registered owner of the current token index
-        let (high, low) = split_felt(index);
-        let index_uint256 = Uint256(low=low, high=high);
-        let (token_id) = IERC721Enumerable.tokenByIndex(
-            contract_address=contract_address, index=index_uint256
+        // [Check] Contract token balance is null then return zero
+        let (contract_address) = get_contract_address();
+        let (carbonable_project_address) = CarbonableOffseter_carbonable_project_address_.read();
+        let (balance) = IERC721.balanceOf(
+            contract_address=carbonable_project_address, owner=contract_address
         );
-        let (owner) = CarbonableOffseter_registered_owner_.read(token_id);
-
-        // Check if user is registered owner
-        let not_eq = is_not_zero(owner - user);
-        if (not_eq == TRUE) {
-            // Stop if index = 0
-            if (index == 0) {
-                return (tokens_len=tokens_len, tokens=tokens);
-            }
-
-            // Else move on to next index
-            let (updated_tokens_len, updated_tokens) = _registerd_tokens_iter(
-                contract_address=contract_address,
-                user=user,
-                index=index - 1,
-                tokens_len=tokens_len,
-                tokens=tokens,
-            );
-            return (tokens_len=updated_tokens_len, tokens=updated_tokens);
+        let zero = Uint256(low=0, high=0);
+        let (equal) = uint256_eq(balance, zero);
+        if (equal == TRUE) {
+            return (token_id=zero);
         }
 
-        // User is the registered owner, the add the token id to the tokens array
-        assert tokens[tokens_len] = token_id;
-
-        // Stop if index is null
-        if (index == 0) {
-            return (tokens_len=tokens_len + 1, tokens=tokens);
-        }
-        let (updated_tokens_len, updated_tokens) = _registerd_tokens_iter(
-            contract_address=contract_address,
-            user=user,
-            index=index - 1,
-            tokens_len=tokens_len + 1,
-            tokens=tokens,
+        // [Check] Otherwise return token id
+        let (token_id) = IERC721Enumerable.tokenOfOwnerByIndex(
+            contract_address=carbonable_project_address, owner=contract_address, index=zero
         );
-        return (tokens_len=updated_tokens_len, tokens=updated_tokens);
+
+        return (token_id=token_id);
     }
 
     func _total_claimed_iter{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
```

Comment:
Curious to know how much steps does this take VS a tail-recursive version

Reply:
Done